Bump coroutines version to 1.6

malinajirka opened the pull request:
Bumps coroutinesVersion from 1.5.2 to 1.6.1.

[Updates kotlinx-coroutines-core from 1.5.2 to 1.6.1](https://github.com/Kotlin/kotlinx.coroutines/blob/master/CHANGES.md)

[Updates kotlinx-coroutines-android from 1.5.2 to 1.6.1](https://github.com/Kotlin/kotlinx.coroutines/blob/master/CHANGES.md)

[Updates kotlinx-coroutines-test from 1.5.2 to 1.6.1](https://github.com/Kotlin/kotlinx.coroutines/releases) + [migration guide](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-test/MIGRATION.md)


I was not able to migrate ReviewModerationHandlerTests.kt. I tried several different approaches using both Standard and Unconfined dispatchers, but at least 3 tests were always failing. I'm probably missing something or the tests need to be rewritten. I decided to give up for now, since the tests work with the original approach.

Again, it is recommended to review this commit by commit to follow my train of thoughts. I'm sorry for the number of changes - I didn't realize that tests which live in the CardReaderModule are not using any BaseUnitTest class. Let me know if you'd prefer me to split it into two PRs.

dependabot[bot] committed:
Bump coroutinesVersion from 1.5.2 to 1.6.1

Bumps `coroutinesVersion` from 1.5.2 to 1.6.1.

Updates `kotlinx-coroutines-core` from 1.5.2 to 1.6.1
- [Release notes](https://github.com/Kotlin/kotlinx.coroutines/releases)
- [Changelog](https://github.com/Kotlin/kotlinx.coroutines/blob/master/CHANGES.md)
- [Commits](https://github.com/Kotlin/kotlinx.coroutines/compare/1.5.2...1.6.1)

Updates `kotlinx-coroutines-android` from 1.5.2 to 1.6.1
- [Release notes](https://github.com/Kotlin/kotlinx.coroutines/releases)
- [Changelog](https://github.com/Kotlin/kotlinx.coroutines/blob/master/CHANGES.md)
- [Commits](https://github.com/Kotlin/kotlinx.coroutines/compare/1.5.2...1.6.1)

Updates `kotlinx-coroutines-test` from 1.5.2 to 1.6.1
- [Release notes](https://github.com/Kotlin/kotlinx.coroutines/releases)
- [Changelog](https://github.com/Kotlin/kotlinx.coroutines/blob/master/CHANGES.md)
- [Commits](https://github.com/Kotlin/kotlinx.coroutines/compare/1.5.2...1.6.1)

---
updated-dependencies:
- dependency-name: org.jetbrains.kotlinx:kotlinx-coroutines-core
  dependency-type: direct:production
  update-type: version-update:semver-minor
- dependency-name: org.jetbrains.kotlinx:kotlinx-coroutines-android
  dependency-type: direct:production
  update-type: version-update:semver-minor
- dependency-name: org.jetbrains.kotlinx:kotlinx-coroutines-test
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

malinajirka committed:
Update advanceXyz imports

malinajirka committed:
Use runTest and UnconfinedTestDispatcher

malinajirka committed:
Replace TestCoroutineScope with TestScope

malinajirka committed:
Minor fixes

malinajirka committed:
Update failing tests

malinajirka committed:
Introduce BaseUnitTest in CardReaderModule

malinajirka committed:
Use runTest in CardReaderModule

malinajirka committed:
Remove unnecessary stubbings

malinajirka committed:
Fix CardReaderPaymentViewModelTests

malinajirka committed:
Revert ReviewModerationHandlerTests to deprecated approach

malinajirka committed:
Fix detekt

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/variations/OrderCreationVariationSelectionViewModelTest.kt:
```diff
@@ -87,7 +91,7 @@ class OrderCreationVariationSelectionViewModelTest : BaseUnitTest() {
                 if (index == 0) productVariation.copy(price = null) else productVariation
             }
-        whenever(variationsRepository.fetchProductVariations(PRODUCT_ID))
+        whenever(variationsRepository.getProductVariationList(PRODUCT_ID))
```
ℹ️ It seems the goal of this test is to check that variations without price are filtered out.

The order in which the coroutines are processed in this test has changed and `viewModel.viewState.getOrAwaitValue()` returns the first value which is set - that's 5 items [with a price](https://github.com/woocommerce/woocommerce-android/blob/ed98cecfb7b278cd4298a2e8c5aebe319e7a9602/WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/variations/OrderCreationVariationSelectionViewModelTest.kt#L33). This PR changes the mocked value and returns a list with one item without a price for  `getProductVariationList` to fix the test.

hichamboushaba replied:
I don't think the right fix is to change the mocked function, as if you check the implementation, we start by emitting the cached values, then we fetch the updated values from network, which means these last values are the ones we care about.

I think after changing the scheduler, the `asLiveData` behavior changed too (as it uses the Main.immediate dispatcher), and now it receives the two values explicitly, where before it probably did some conflating, and received only the last value. With this in mind, the `getOrAwaitValue` doesn't work anymore, as it will just return the first element.

So to fix the test, without changing its behavior, we can do something like this:
```kotlin
    @Test
    fun `when loading variations, then exclude any variations without price`() = testBlocking {
        val variations = ProductTestUtils.generateProductVariationList(PRODUCT_ID)
            .mapIndexed { index, productVariation ->
                if (index == 0) productVariation.copy(price = null) else productVariation
            }
        whenever(variationsRepository.fetchProductVariations(PRODUCT_ID))
            .thenReturn(variations)

        var displayedVariations: List<ProductVariation>? = null
        viewModel.viewState.observeForever {
            displayedVariations = it.variationsList
        }
        
        assertThat(displayedVariations).allMatch { it.price != null }
        assertThat(displayedVariations?.size).isEqualTo(variations.size - 1)
    }
```
WDYT?

malinajirka replied:
Thanks for sharing your thoughts and the code 🙇 !

>  think after changing the scheduler, the asLiveData behavior changed too (as it uses the Main.immediate dispatcher), and now it receives the two values explicitly, where before it probably did some conflating, and received only the last value. With this in mind, the getOrAwaitValue doesn't work anymore, as it will just return the first element.

My train of thoughts was the same.

> I don't think the right fix is to change the mocked function, as if you check the implementation, we start by emitting the cached values, then we fetch the updated values from network, which means these last values are the ones we care about.

Yes, but it seems the goal of the test is to check that items are filtered not that first a cached value is emitted and then a data from the remote are emitted -> that's why I opted for changing the test, it imho also better corresponds with the test name which says "loading variations" not "fetching variations".

Anyway, if you agree I opted for another solution - I added the code you shared as a new test. So we check both cases and are 100% sure it works as expected. ea761d037e3276f10c6efdf48921212088bd7f35

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt:
```diff
@@ -23,7 +30,7 @@ import org.wordpress.android.fluxc.network.rest.wpcom.wc.WooErrorType.GENERIC_ER
 import org.wordpress.android.fluxc.store.WCProductStore
 
-class ReviewModerationHandlerTests : BaseUnitTest() {
+class ReviewModerationHandlerTests {
```
I wasn't able to make this test suite work with the new approach - decided to revert to the original.

hichamboushaba replied:
I can work on this in a separate PR.

I think there are two reasons for issues with the tests on this class:
1. The main reason is the use of `UnconfinedTestScheduler` as parameter for `runTest`, as it's impossible to dictate the order of execution when using it, where this test suite depends on the order a lot.
2. The second reason is not being able to collect all items without conflating from the exposed states flow, and this is also related to the used Dispatcher (there are a some tickets about this in the Github repo of the library: https://github.com/Kotlin/kotlinx.coroutines/issues/3120, as it's a bit confusing)

malinajirka replied:
> The main reason is the use of UnconfinedTestScheduler as parameter for runTest, as it's impossible to dictate the order of execution when using it, where this test suite depends on the order a lot.

I tried making it work even with the StandardTestDispatcher which uses the standardScheduler (even with a combination of the dispatchers). However, I think one vital mistake I did was that I forgot that there were two similar methods runTestAndReturnLastEmittedStatusList and runTestAndCollectAllStatuses - n/c 😞 .

> I can work on this in a separate PR

That'd be awesome, thanks so much! Or we could do a pair programming session on this, I think it might be valuable for me :).

wzieba commented:
@malinajirka if it's something you'd like to try, #6251 (hopefully) fixed CodeCov integration. I think it could work if you'll merge the current `trunk` 🙂.

hichamboushaba commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/CreateOrUpdateOrderDraftTests.kt:
```diff
@@ -82,6 +85,7 @@ class CreateOrUpdateOrderDraftTests : BaseUnitTest() {
 
         verify(orderCreationRepository, never()).createOrUpdateDraft(any())
         advanceTimeBy(CreateOrUpdateOrderDraft.DEBOUNCE_DURATION_MS)
+        runCurrent()
```
I wonder if an extension like the one below can be useful to avoid having to make calls like this with the new scheduler behavior?
```kotlin
fun TestScope.advanceTimeAndRun(duration: Long) {
        advanceTimeBy(duration)
        runCurrent()
}
```

And an np comment, I think the deleted empty line should be restored to separate the "act" part from the "assert" one.

malinajirka replied:
> I wonder if an extension like the one below can be useful to avoid having to make calls like this with the new scheduler behavior?

Great idea, added in 829934809b526e43a15b63669e5b38eddef96889 and d606d828aadfb791ad4bf323b8f93adb1572f8a7

> And an np comment, I think the deleted empty line should be restored to separate the "act" part from the "assert" one.

I'm not sure which deleted empty line you mean - I don't see any such change in this class. Either way, feel free to update the code or let me know and I'll update it.

hichamboushaba replied:
> I'm not sure which deleted empty line you mean

I must've been looking at different line while writing the comment 😅 🤦, please ignore this part.

hichamboushaba commented on WooCommerce/src/test/kotlin/com/woocommerce/android/util/CoroutineTestRule.kt:
```diff
@@ -2,14 +2,15 @@ package com.woocommerce.android.util
 
 @ExperimentalCoroutinesApi
-class CoroutineTestRule(val testDispatcher: TestCoroutineDispatcher = TestCoroutineDispatcher()) : TestWatcher() {
+class CoroutineTestRule(val testDispatcher: TestDispatcher = UnconfinedTestDispatcher()) : TestWatcher() {
```
A suggestion: while I agree that using `UnconfinedTestDispatcher` is a good default for most of our test cases, it shouldn't be the only way, some tests require some definite and explicit ordering (like the `ReviewModerationHandlerTests` above), and those would benefit from using `StandardTestDispatcher`.
So I suggest removing the default value from here, and adding an argument to specify the dispatcher in `BaseUnitTest` with `UnconfinedTestDispatcher` as a default value, this would allow child test classes to change the dispatcher easily.

malinajirka replied:
Great suggestion! I actually had this implemented while trying to fix the reviewHandler tests but reverted it with all the other changes I made. Fixed in f6662b2fe035eb6dba10b6a42eb6c12b3fbeb1be

hichamboushaba commented on libs/cardreader/src/test/java/com/woocommerce/android/cardreader/internal/CardReaderBaseUnitTest.kt:
```diff
@@ -0,0 +1,22 @@
+
+@Suppress("UnnecessaryAbstractClass")
+@RunWith(MockitoJUnitRunner::class)
+abstract class CardReaderBaseUnitTest {
```
I wonder if we shouldn't have a `core` module for things that need to be part of all modules? Maybe we can start with a `core-testing` just for test helpers? Not necessarily on this PR though.

malinajirka replied:
Yeah, I think it'd be great but might need a little more work. Especially when it comes to documentation - we need to make sure we use it correctly and don't add everything that needs to be shared by at least 2 modules there.

hichamboushaba reviewed (approved):
Thanks for working on this @malinajirka, this took me very long to review, because I had to learn about the new APIs before, but it's worth it.
I'm pre-approving the PR, I left some comments, but nothing is blocking, great job.

hichamboushaba commented:
Another point, I think before merging this, you should make sure you are up-to-date with `trunk`, to avoid breaking tests that were added after creating your branch.

malinajirka committed:
Add new test to OrderCreationVariationSelection suite

malinajirka committed:
Introduce advanceTimeAndRun extension method

malinajirka committed:
Add support for setting a custom dispatcher

malinajirka committed:
Merge branch 'trunk' into bump-coroutines-version-to-1.6

malinajirka committed:
Move CoroutineTestUtils to unit test package

malinajirka commented:
Thanks so much for the great review @hichamboushaba! Truly appreciate the time you spent on this 🙇.

I've implemented most of the suggestions you proposed, great tips! I've also merged trunk into the branch and double checked no new tests which would be affected by the changes were added. Essentially, the only thing I didn't implement is creating a separate "core" module. 

I think it'd be great if you checked the commits I added before we merge this just to be sure I didn't mess anything up. I'll update it with trunk one more time before the final merge, let me know. Thanks so much!!!

Btw I forgot to mention it in the description - this PR changes not just the test package but also the core coroutine package -> if you haven't tested the app it might be worth testing. I run some tests and haven't noticed anything weird.

malinajirka committed:
Fix detekt

malinajirka committed:
Add missing import

hichamboushaba reviewed (approved):
Looks great, thanks for the changes @malinajirka :shipit:

malinajirka committed:
Merge branch 'trunk' into bump-coroutines-version-to-1.6
